test/unit/org/apache/cassandra/tools/ToolRunner.java

```diff
@@ -0,0 +1,432 @@
+import org.slf4j.LoggerFactory;
+
+import org.apache.cassandra.cql3.CQLTester;
+import org.apache.cassandra.tools.OfflineToolUtils.SystemExitException;
+
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.fail;
+
+public class ToolRunner implements AutoCloseable
+{
+    protected static final Logger logger = LoggerFactory.getLogger(ToolRunner.class);
+    
+    private final List<String> allArgs = new ArrayList<>();
+    private Process process;
+    private final ByteArrayOutputStream errBuffer = new ByteArrayOutputStream();
+    private final ByteArrayOutputStream outBuffer = new ByteArrayOutputStream();
+    private InputStream stdin;
+    private boolean stdinAutoClose;
+    private long defaultTimeoutMillis = TimeUnit.SECONDS.toMillis(30);
+    private Thread ioWatcher;
+    private Map<String, String> envs;
+    private boolean runOutOfProcess = true;
+
+    public ToolRunner(List<String> args)
+    {
+        this.allArgs.addAll(args);
+    }
+    
+    public ToolRunner(List<String> args, boolean runOutOfProcess)
+    {
+        this.allArgs.addAll(args);
+        this.runOutOfProcess = runOutOfProcess;
+    }
+
+    public ToolRunner withStdin(InputStream stdin, boolean autoClose)
+    {
+        this.stdin = stdin;
+        this.stdinAutoClose = autoClose;
+        return this;
+    }
+
+    public ToolRunner withEnvs(Map<String, String> envs)
+    {
+        Preconditions.checkArgument(runOutOfProcess, "Not supported");
+        this.envs = envs;
+        return this;
+    }
+
+    public ToolRunner start()
+    {
+        if (process != null)
+            throw new IllegalStateException("Process already started. Create a new ToolRunner instance for each invocation.");
+
+        logger.debug("Starting {} with args {}", runOutOfProcess ? "process" : "class" , argsToLogString());
+
+        try
+        {
+            if (runOutOfProcess)
+            {
+                ProcessBuilder pb = new ProcessBuilder(allArgs);
+                if (envs != null)
+                    pb.environment().putAll(envs);
+                process = pb.start();
+            }
+            else
+            {
+                PrintStream originalSysOut = System.out;
+                PrintStream originalSysErr = System.err;
+                InputStream originalSysIn = System.in;
+                originalSysOut.flush();
+                originalSysErr.flush();
+                ByteArrayOutputStream toolOut = new ByteArrayOutputStream();
+                ByteArrayOutputStream toolErr = new ByteArrayOutputStream();
+                System.setOut(new PrintStream(toolOut));
+                System.setErr(new PrintStream(toolErr));
+                System.setIn(stdin == null ? originalSysIn : stdin);
+                
+                String clazz = allArgs.get(0);
+                String[] clazzArgs = allArgs.subList(1, allArgs.size()).toArray(new String[0]);
+                int exitCode = runClassAsTool(clazz, clazzArgs);
```

Comment:
Better to explicitly close the 2 `PrintStream` instances created above using try-finally around `runClassAsTool`, because `PrintStream` can buffer data.

Reply:
+1